alembic/script/revision.py

```diff
@@ -886,6 +895,122 @@ def _iterate_related_revisions(self, fn, targets, map_, check=False):
                         )
                     )
 
+    def topological_sort(self, revisions, reverse=False):
+        """Yield revision ids of a collection of Revision objects in
+        topological sorted order (i.e. revisions always come after their
+        down_revisions and dependencies)."""
+        allitems = [d.revision for d in revisions]
+        edges = [
+            (rev, child.revision)
+            for child in revisions
+            if child.down_revision is not None
+            for rev in util.to_tuple(child.down_revision)
+        ] + [
+            (parent, child.revision)
+            for child in revisions
+            if child.dependencies is not None
+            for parent in util.to_tuple(child.dependencies)
+        ]
+        return sqlautil.topological.sort(edges, allitems)
+
+    def walk_down(self, start, steps, label):
+        """ Walk down the tree along a single path. """
+        assert steps <= 0
+        if isinstance(start, compat.string_types):
+            start = self.get_revision(start)
+        if steps == 0:
+            return start
+        if start is None:
+            raise ValueError("Can't walk back past base.")
+        children = self.get_revisions(start.down_revision)
+        if len(children) == 0:
+            raise util.CommandError(
+                f"Relative revision {label} didn't produce "
+                f"{abs(int(label))} migrations"
+            )
+        assert len(children) == 1
+        return self.walk_down(next(iter(children)), steps + 1, label)
+
+    def walk(self, start, steps, branch_label):
+        """ Walk up the tree along a single path. """
+        assert steps >= 0
+        if isinstance(start, compat.string_types):
+            start = self.get_revision(start)
+        if steps == 0:
+            return start
+        if start is None:
+            roots = [
+                rev
+                for rev in self._revision_map.values()
+                if rev is not None and rev.down_revision is None
+            ]
+            assert len(roots) == 1
+            return self.walk(roots[0], steps - 1, branch_label)
+        children = [
+            rev
+            for rev in self.get_revisions(start.nextrev)
+            if branch_label is None or branch_label in rev.branch_labels
+        ]
+        assert len(children) == 1
+        return self.walk(next(iter(children)), steps - 1, branch_label)
```

Comment:
**mike bayer** (zzzeek) wrote:

same idea here...

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530

Reply:
**Simon Bowly** (simonbowly) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530